Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -613,15 +641,25 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
 
         // We don't need the cumulation buffer, if we're holding it.
         self.cumulationBuffer = nil
-
+        
+        self.firePendingInOut(ctx: ctx)
+        
         // No check to state.currentError because, if we hit it before, we already threw that
         // error. This never calls any of the callbacks that set that field anyway. Instead we
         // just check if the errno is set and throw.
+        if let parserError = self.currentParserError() {
+            self.state.currentError = parserError
+            ctx.fireErrorCaught(parserError)
+        }
+    }
+    
+    private func currentParserError() -> HTTPParserError? {
         let httpError = self.parser.http_errno
-        if httpError != 0 {
-            self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
-            ctx.fireErrorCaught(self.state.currentError!)
+        // Also take into account that we may have called c_nio_http_parser_pause(...)
+        guard httpError != 0 && httpError != HPE_PAUSED.rawValue else {
```

Comment:
@weissi I guess it not really matters much for perf stuff but we may want to consider switching these as most of the time it will be HPE_PAUSED in reality so the first test is a waste mostly.

Reply:
reordered this one